piptools/repositories/local.py

```diff
@@ -76,10 +77,9 @@ def find_best_match(
         key = key_from_ireq(ireq)
         existing_pin = self.existing_pins.get(key)
         if existing_pin and ireq_satisfied_by_existing_pin(ireq, existing_pin):
-            project, version, _ = as_tuple(existing_pin)
-            return make_install_requirement(project, version, ireq)
-        else:
-            return self.repository.find_best_match(ireq, prereleases)
+            with suppress(NoCandidateFound):
+                return self.repository.find_best_match(existing_pin, prereleases)
```

Comment:
Also note this would probably slow down the re-compile process because it would always hit index_urls and find_links.

Reply:
Hmm, well it seems to me that in order to do the "right thing" here (avoid a hard failure on bad already-compiled versions), we *need* to consult the index about the already-compiled versions.

Reply:
The question is what's more important "speed" or "right thing, but in rare case"?

Reply:
I did some simple benchmark.

Before the fix:
```bash
$ echo web3 > requirements.in

# first run
$ time pip-compile
real	0m18.898s
user	0m11.924s
sys	0m0.993s

# second run
$ time pip-compile
real	0m0.589s
user	0m0.517s
sys	0m0.072s
```


After the fix:
```bash
$ echo web3 > requirements.in

# first run
$ time pip-compile
real	0m17.062s
user	0m12.388s
sys	0m0.970s

# second run
$ time pip-compile
real	0m3.631s
user	0m2.583s
sys	0m0.120s
```

As I expected re-compile would be slower after the fix (6 times slower on my benchmark) due to the hitting to the simple API for every compiled package.

I understand the intention to do the right thing, but the issue doesn't seem critical and slows down pip-compile on re-runs, which might not be good. Also, the legacy resolver will be removed soon, and spending much energy on it might not be worth it.

What do you think if we only merge the test which I find pretty helpful?

Reply:
Thanks for running the numbers!

> What do you think if we only merge the test which I find pretty helpful?

Sure, sounds good to me.

Reply:
Thank you! Would you like to open a separate PR?

Reply:
Sure, let's continue at #1774